Master restart fail

**Reproduce Step**

1. start master, start executor, submit a job
2. stop the master and executor(kill -9)
3. restart the master 

**Expect**

Master log some errors, like `no executor` or resources

**Acutal**

Master exit and panic
 
**Related log**

[2022/03/15 12:15:19.857 +08:00] [INFO] [job_fsm.go:189] ["job master doesn't receive heartbeat in time, recreate it"] [job="{\"WorkerHandle\":null,\"id\":\"fbf59cc5-189b-41cc-b3c9-20035396e648\",\"addr\":\"127.0.0.1:10241\",\"node-id\":\"1c431ca9-6dea-4dc2-9f9f-c65bf25ac963\",\"epoch\":10,\"status\":1,\"type\":0,\"config\":null}"]
[2022/03/15 12:15:19.857 +08:00] [INFO] [server.go:771] [payload="tasks:<task:<> cost:1 > "] [request=ScheduleTask]
[2022/03/15 12:15:19.858 +08:00] [WARN] [jobmanager.go:261] ["dispatch worker met error"] [error="rpc error: code = Unknown desc = [DFLOW:ErrClusterResourceNotEnough]cluster resource is not enough, please scale out the cluster"] [errorVerbose="rpc error: code = Unknown desc = [DFLOW:ErrClusterResourceNotEnough]cluster resource is not enough, please scale out the cluster\ngithub.com/pingcap/errors.AddStack\n\t/Users/maxshuang/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.Trace\n\t/Users/maxshuang/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/juju_adaptor.go:15\ngithub.com/hanfei1991/microcosm/lib.(*DefaultBaseMaster).CreateWorker.func1\n\t/Users/maxshuang/GolandProjects/dataflow/microcosm/lib/master.go:565\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1133"]
[2022/03/15 12:15:20.261 +08:00] [ERROR] [campaign.go:94] ["run leader service failed"] [error="[DFLOW:ErrWorkerNotFound]worker is not found: worker ID 43fa8521-8fb8-4da7-9154-3e88b1715724"] [errorVerbose="[DFLOW:ErrWorkerNotFound]worker is not found: worker ID 43fa8521-8fb8-4da7-9154-3e88b1715724\ngithub.com/pingcap/errors.AddStack\n\t/Users/maxshuang/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\t/Users/maxshuang/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/normalize.go:164\ngithub.com/hanfei1991/microcosm/servermaster.(*JobFsm).JobDispatchFailed\n\t/Users/maxshuang/GolandProjects/dataflow/microcosm/servermaster/job_fsm.go:233\ngithub.com/hanfei1991/microcosm/servermaster.(*JobManagerImplV2).OnWorkerDispatched\n\t/Users/maxshuang/GolandProjects/dataflow/microcosm/servermaster/jobmanager.go:262\ngithub.com/hanfei1991/microcosm/lib.(*DefaultBaseMaster).CreateWorker.func1\n\t/Users/maxshuang/GolandProjects/dataflow/microcosm/lib/master.go:565\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1133"]
[2022/03/15 12:15:20.333 +08:00] [INFO] [campaign.go:105] ["start to campaign server master leader"] [name=server-master-1-03abb9be-37c6-4181-bdf7-ee423ba37613]
[2022/03/15 12:15:20.352 +08:00] [INFO] [campaign.go:117] ["campaign leader successfully"] [name=server-master-1-03abb9be-37c6-4181-bdf7-ee423ba37613]
[2022/03/15 12:15:20.352 +08:00] [INFO] [executor_manager.go:124] ["register executor"] [info="{\"type\":2,\"id\":\"1c431ca9-6dea-4dc2-9f9f-c65bf25ac963\",\"addr\":\"127.0.0.1:10241\",\"cap\":100}"]
[2022/03/15 12:15:20.352 +08:00] [INFO] [capacity_impl.go:39] ["executor resource is registered"] [executor-id=1c431ca9-6dea-4dc2-9f9f-c65bf25ac963] [capacity=100]
[2022/03/15 12:15:20.352 +08:00] [INFO] [master_client.go:100] ["add new server master client"] [addr=0.0.0.0:10240]
[2022/03/15 12:15:20.369 +08:00] [WARN] [metadata.go:194] ["master meta exits, will be overwritten"] [old-meta="{\"id\":\"dataflow-engine-job-manager\",\"addr\":\"\",\"node-id\":\"\",\"epoch\":9,\"status\":2,\"type\":1,\"config\":null}"]
[2022/03/15 12:15:20.376 +08:00] [PANIC] [server.go:498] ["duplicate handlers"] [topic=heartbeat-ping-dataflow-engine-job-manager] [stack="github.com/pingcap/tiflow/pkg/p2p.(*MessageServer).registerHandler\n\t/Users/maxshuang/go/pkg/mod/github.com/pingcap/tiflow@v0.0.0-20220308162507-e342c02e789c/pkg/p2p/server.go:498\ngithub.com/pingcap/tiflow/pkg/p2p.(*MessageServer).run\n\t/Users/maxshuang/go/pkg/mod/github.com/pingcap/tiflow@v0.0.0-20220308162507-e342c02e789c/pkg/p2p/server.go:231\ngithub.com/pingcap/tiflow/pkg/p2p.(*MessageServer).Run.func2\n\t/Users/maxshuang/go/pkg/mod/github.com/pingcap/tiflow@v0.0.0-20220308162507-e342c02e789c/pkg/p2p/server.go:194\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/Users/maxshuang/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57"]
panic: duplicate handlers

goroutine 277 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0x1400076a0c0, {0x14001614080, 0x1, 0x1})
	/Users/maxshuang/go/pkg/mod/go.uber.org/zap@v1.20.0/zapcore/entry.go:232 +0x3c8
go.uber.org/zap.(*Logger).Panic(0x14001634120, {0x105bd6cd8, 0x12}, {0x14001614080, 0x1, 0x1})
	/Users/maxshuang/go/pkg/mod/go.uber.org/zap@v1.20.0/logger.go:230 +0x68
github.com/pingcap/log.Panic({0x105bd6cd8, 0x12}, {0x14001614080, 0x1, 0x1})
	/Users/maxshuang/go/pkg/mod/github.com/pingcap/log@v0.0.0-20211215031037-e024ba4eb0ee/global.go:54 +0x100
github.com/pingcap/tiflow/pkg/p2p.(*MessageServer).registerHandler(0x14000614380, {0x10649d550, 0x140012e4000}, {0x14000124fc0, 0x2a}, {0x1064a93f0, 0x140003c88c0}, 0x140007705a0)
	/Users/maxshuang/go/pkg/mod/github.com/pingcap/tiflow@v0.0.0-20220308162507-e342c02e789c/pkg/p2p/server.go:498 +0x144
github.com/pingcap/tiflow/pkg/p2p.(*MessageServer).run(0x14000614380, {0x10649d550, 0x140012e4000})
	/Users/maxshuang/go/pkg/mod/github.com/pingcap/tiflow@v0.0.0-20220308162507-e342c02e789c/pkg/p2p/server.go:231 +0x36c
github.com/pingcap/tiflow/pkg/p2p.(*MessageServer).Run.func2()
	/Users/maxshuang/go/pkg/mod/github.com/pingcap/tiflow@v0.0.0-20220308162507-e342c02e789c/pkg/p2p/server.go:194 +0x38
golang.org/x/sync/errgroup.(*Group).Go.func1(0x140012e6000, 0x1400097f4a0)
	/Users/maxshuang/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go
	/Users/maxshuang/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x60

--- a/servermaster/server.go
+++ b/servermaster/server.go
@@ -611,6 +611,10 @@ func (s *Server) runLeaderService(ctx context.Context) (err error) {
 	defer func() {
 		s.leader.Store(&Member{})
 		s.resign()
+		err := s.jobManager.Close(ctx)
+		if err != nil {
+			log.L().Warn("job manager close with error", zap.Error(err))
+		}
 	}()
 
 	metricTicker := time.NewTicker(defaultMetricInterval)

--- a/servermaster/server_test.go
+++ b/servermaster/server_test.go
@@ -226,6 +226,14 @@ func TestRunLeaderService(t *testing.T) {
 	err = s.runLeaderService(ctx1)
 	require.EqualError(t, err, context.DeadlineExceeded.Error())
 
+	// runLeaderService exits, try to campaign to be leader and run leader servcie again
+	err = s.campaign(ctx, time.Second)
+	require.Nil(t, err)
+	ctx2, cancel2 := context.WithTimeout(ctx, time.Second)
+	defer cancel2()
+	err = s.runLeaderService(ctx2)
+	require.EqualError(t, err, context.DeadlineExceeded.Error())
+
 	cancel()
 	wg.Wait()
 }

